server/src/main/java/org/elasticsearch/cluster/coordination/PublicationTransportHandler.java

```diff
@@ -197,20 +200,41 @@ private PublishWithJoinResponse handleIncomingPublishRequest(BytesTransportReque
                         incomingState.stateUUID(),
                         request.bytes().length()
                     );
-                    final PublishWithJoinResponse response = acceptState(incomingState);
-                    lastSeenClusterState.compareAndSet(lastSeen, incomingState);
-                    return response;
+                    acceptState(
+                        incomingState,
+                        transportChannel,
+                        (acceptedState) -> lastSeenClusterState.compareAndSet(lastSeen, acceptedState)
+                    );
                 }
             }
         } finally {
             IOUtils.close(in);
         }
     }
 
-    private PublishWithJoinResponse acceptState(ClusterState incomingState) {
+    /**
+     * Delegate to cluster-coordination thread to apply received state
+     *
+     * @param incomingState The received cluster state
+     * @param transportChannel The channel on which it was received
+     * @param onSuccess The action to perform if the new state is accepted
+     */
+    private void acceptState(ClusterState incomingState, TransportChannel transportChannel, Consumer<ClusterState> onSuccess) {
```

Comment:
I think I'd rather we avoided passing the `TransportChannel` directly here. It works, but it's more idiomatic to have async methods accept an `ActionListener<...>` - see the Javadocs for `org.elasticsearch.action.ActionListener` for lots more detail, and also it's useful to constrain the response type as early as possible. In this case a `org.elasticsearch.action.support.ChannelActionListener<PublishWithJoinResponse>` would be a good choice. I think the same comment applies to `handleIncomingPublishRequest` - on the one hand, using a method reference here is kinda neat, but on the other hand we don't use the `Task` argument and it would be easier on the reader to pass in an `ActionListener<PublishWithJoinResponse>` here too.

Reply:
Yes that looks much nicer, thanks!